network_analysis/trace-a-utility-network/src/main/java/com/esri/samples/trace_a_utility_network/TraceAUtilityNetworkController.java

```diff
@@ -0,0 +1,516 @@
+                  }
+                }
+              }
+            }
+          }
+        } catch (InterruptedException | ExecutionException ex) {
+          statusLabel.setText("Error identifying clicked features.");
+          new Alert(Alert.AlertType.ERROR, "Error identifying clicked features.").show();
+        } finally {
+          progressIndicator.setVisible(false);
+        }
+      });
+    }
+  }
+
+  /**
+   * Uses a UtilityNetworkSource to create a UtilityElement object out of an ArcGISFeature.
+   *
+   * @param identifiedFeature an ArcGISFeature object that will be used to create a UtilityElement
+   * @param networkSource the UtilityNetworkSource to which the created UtilityElement is associated
+   * @return the created UtilityElement
+   */
+  private UtilityElement createUtilityElement(ArcGISFeature identifiedFeature, UtilityNetworkSource networkSource) {
+    UtilityElement utilityElement = null;
+
+    // get the attributes of the identified feature
+    Map<String, Object> attributes = identifiedFeature.getAttributes();
+
+    // get the name of the utility asset group's attribute field from the feature
+    String assetGroupFieldName = identifiedFeature.getFeatureTable().getSubtypeField();
+
+    // find the code matching the asset group name in the feature's attributes
+    int assetGroupCode = (int) attributes.get(assetGroupFieldName.toLowerCase());
+
+    // iterate through the network source's asset groups to find the group with the matching code
+    List<UtilityAssetGroup> assetGroups = networkSource.getAssetGroups();
+    for (UtilityAssetGroup assetGroup : assetGroups) {
+      if (assetGroup.getCode() == assetGroupCode) {
+
+        // get the code for the feature's asset type from its attributes
+        String assetTypeCode = attributes.get("assettype").toString();
+
+        // iterate through the asset group's asset types to find the type matching the feature's asset type code
+        List<UtilityAssetType> utilityAssetTypes = assetGroup.getAssetTypes();
+        for (UtilityAssetType assetType : utilityAssetTypes) {
+          if (assetType.getCode() == Integer.parseInt(assetTypeCode)) {
+
+            // get the list of terminals for the feature
+            List<UtilityTerminal> terminals = assetType.getTerminalConfiguration().getTerminals();
+
+            // if there is only one terminal, use it to create a utility element
+            if (terminals.size() == 1) {
+              utilityElement = utilityNetwork.createElement(identifiedFeature, terminals.get(0));
+              // show the name of the terminal in the status label
+              showTerminalNameInStatusLabel(terminals.get(0));
+
+              // if there is more than one terminal, prompt the user to select one
+            } else if (terminals.size() > 1) {
+              // create a dialog for terminal selection
+              UtilityTerminalSelectionDialog utilityTerminalSelectionDialog =
```

Comment:
Instead of creating a custom dialog, you could do:

```java
ChoiceDialog<UtilityTerminal> utilityTerminalSelectionDialog = new ChoiceDialog<>(terminals.get(0), terminals);
utilityTerminalSelectionDialog.setTitle("Select Utility Terminal:");
              
// override the list cell in the dialog's combo box to show the terminal name
@SuppressWarnings("unchecked")
ComboBox<UtilityTerminal> comboBox = (ComboBox<UtilityTerminal>) ((GridPane) utilityTerminalSelectionDialog.getDialogPane().getContent()).getChildren().get(1);
comboBox.setCellFactory(param -> new UtilityTerminalListCell());
comboBox.setButtonCell(new UtilityTerminalListCell());
```

I admit it's a bit of a hack to work around `UtilityTerminal.toString` not being helpful, but I think it's a little better than creating a whole dialog for it.

Here's the `ListCell`:
```java
public class UtilityTerminalListCell extends ListCell<UtilityTerminal> {
    @Override
    protected void updateItem(UtilityTerminal item, boolean empty) {
        super.updateItem(item, empty);
        if (empty) {
            setText(null);
        } else {
            setText(item.getName());
        }
    }
}
```

Reply:
I suppose it is one less class so it keeps the focus on the main point of the sample, so happy to implement this!